File: app/src/main/java/com/mapmyindia/sdk/demo/java/adapter/GeoFenceDetailAdapter.java
package com.mapmyindia.sdk.demo.java.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.SwitchCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.mapmyindia.sdk.demo.R;
import com.mapmyindia.sdk.demo.java.model.GeofenceDetail;

import java.util.List;

public class GeoFenceDetailAdapter extends RecyclerView.Adapter<GeoFenceDetailAdapter.ViewHolder> {

    private List<GeofenceDetail> geofenceDetailList;
    private OnGeoFenceChangeListener onGeoFenceChangeListener;

    public void setGeofenceDetailList(List<GeofenceDetail> geofenceDetailList) {
        this.geofenceDetailList = geofenceDetailList;
        notifyDataSetChanged();
    }

    public void setOnGeoFenceChangeListener(OnGeoFenceChangeListener onGeoFenceChangeListener) {
        this.onGeoFenceChangeListener = onGeoFenceChangeListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_geofence_detail_adapter, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        GeofenceDetail geofenceDetail = geofenceDetailList.get(position);
        holder.tvLabel.setText(geofenceDetail.getGfLabel());
        holder.swStatus.setChecked(geofenceDetail.getActive());
        holder.ivEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(onGeoFenceChangeListener != null) {
                    onGeoFenceChangeListener.onEditGeoFence(geofenceDetail);
                }
            }
        });
        holder.ivDelete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(onGeoFenceChangeListener != null) {
                    onGeoFenceChangeListener.onRemoveGeofence(geofenceDetail);
                }
            }
        });
        holder.swStatus.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                geofenceDetailList.get(position).setActive(isChecked);
                holder.swStatus.post(new Runnable() {
                    @Override
                    public void run() {
                        notifyDataSetChanged();
                    }
                });
                if(onGeoFenceChangeListener != null) {
                    onGeoFenceChangeListener.onGeoFenceStatusChange();
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return geofenceDetailList == null ?0: geofenceDetailList.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        private TextView tvLabel;
        private SwitchCompat swStatus;
        private ImageView ivEdit, ivDelete;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            tvLabel = itemView.findViewById(R.id.tv_geofence_label);
            swStatus = itemView.findViewById(R.id.sw_status);
            ivEdit = itemView.findViewById(R.id.iv_edit_geofence);
            ivDelete = itemView.findViewById(R.id.iv_delete_geofence);
        }
    }

    public interface OnGeoFenceChangeListener {
        void onGeoFenceStatusChange();
        void onEditGeoFence(GeofenceDetail geofenceDetail);
        void onRemoveGeofence(GeofenceDetail geofenceDetail);
    }
}

File: app/src/main/java/com/mapmyindia/sdk/demo/kotlin/adapter/GeoFenceDetailAdapter.kt
package com.mapmyindia.sdk.demo.kotlin.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.widget.SwitchCompat
import androidx.recyclerview.widget.RecyclerView
import com.mapmyindia.sdk.demo.R
import com.mapmyindia.sdk.demo.kotlin.model.GeofenceDetail

class GeoFenceDetailAdapter : RecyclerView.Adapter<GeoFenceDetailAdapter.ViewHolder>() {
    private var geofenceDetailList: List<GeofenceDetail>? = null
    private var onGeoFenceChangeListener: OnGeoFenceChangeListener? = null


    fun setGeofenceDetailList(geofenceDetailList: List<GeofenceDetail>?) {
        this.geofenceDetailList = geofenceDetailList
        notifyDataSetChanged()
    }

    fun setOnGeoFenceChangeListener(onGeoFenceChangeListener: OnGeoFenceChangeListener?) {
        this.onGeoFenceChangeListener = onGeoFenceChangeListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_geofence_detail_adapter, parent, false)
        return ViewHolder(view)
    }

    override fun getItemCount(): Int {
        return geofenceDetailList?.size?: 0
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val geofenceDetail = geofenceDetailList!![position]
        holder.tvLabel.setText(geofenceDetail.gfLabel)
        holder.swStatus.isChecked = geofenceDetail.active
        holder.ivEdit.setOnClickListener { onGeoFenceChangeListener?.onEditGeoFence(geofenceDetail) }
        holder.ivDelete.setOnClickListener {
            onGeoFenceChangeListener?.onRemoveGeofence(
                    geofenceDetail
            )
        }
        holder.swStatus.setOnCheckedChangeListener { buttonView, isChecked ->
            geofenceDetailList!![position].active = isChecked
            holder.swStatus.post { notifyDataSetChanged() }
            onGeoFenceChangeListener?.onGeoFenceStatusChange()
        }
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val swStatus: SwitchCompat
        val tvLabel: TextView
        val ivEdit: ImageView
        val ivDelete: ImageView

        init {
            swStatus = itemView.findViewById(R.id.sw_status)
            tvLabel = itemView.findViewById(R.id.tv_geofence_label)
            ivEdit = itemView.findViewById(R.id.iv_edit_geofence)
            ivDelete = itemView.findViewById(R.id.iv_delete_geofence)
        }

    }

    interface OnGeoFenceChangeListener {
        fun onGeoFenceStatusChange()
        fun onEditGeoFence(geofenceDetail: GeofenceDetail?)
        fun onRemoveGeofence(geofenceDetail: GeofenceDetail?)
    }
}